kart/diff_structs.py

```diff
@@ -439,16 +446,83 @@ class DeltaDiff(Diff):
     child_type = Delta
 
     def __init__(self, initial_contents=()):
+        self._lazy_initial_contents = None
         if isinstance(initial_contents, (dict, UserDict)):
             super().__init__(initial_contents)
         else:
+            if isinstance(initial_contents, Iterator):
+                self._lazy_initial_contents = (
+                    (delta.key, delta) for delta in initial_contents
+                )
+                initial_contents = ()
             super().__init__((delta.key, delta) for delta in initial_contents)
 
+    def __getitem__(self, key):
+        if key in self.data:
+            return self.data[key]
+        self._evaluate_lazy_initial_contents()
+        return self.data[key]
+
     def __setitem__(self, key, delta):
         if key != delta.key:
             raise ValueError("Delta must be added at the appropriate key")
         super().__setitem__(key, delta)
 
+    def _evaluate_lazy_initial_contents(self):
+        if self._lazy_initial_contents is None:
+            return
+        for k, v in self._lazy_initial_contents:
+            if k not in self:
```

Comment:
Can it happen that k is already in self? Or, can you explain this line?

Reply:
I was thinking about diff concatenation - deltas that are explicitly added to this diff after the fact should probably override the lazy deltas. But in fact that's not correct because those deltas may in fact need to be _combined_ with the lazy deltas.
It looks like this isn't actually required, since diff concatenation calls `.keys()` anyway and thus consumes the generator fully. Which will work correctly.